core/src/main/java/org/elasticsearch/index/mapper/Mapper.java

```diff
@@ -177,4 +177,11 @@ public final String simpleName() {
     /** Return the merge of {@code mergeWith} into this.
      *  Both {@code this} and {@code mergeWith} will be left unmodified. */
     public abstract Mapper merge(Mapper mergeWith, boolean updateAllTypes);
+
+    /**
+     * Update the field type of this mapper. This is necessary because some mapping updates
+     * can modify mappings across several types. This method must return a copy of the mapper
+     * so that the current mapper is not modified.
+     */
+    public abstract Mapper updateFieldType(Map<String, MappedFieldType> fullNameToFieldType);
```

Comment:
I can't disagree with anything you said. But the fact that MappedFieldTypeReference is mutable (by nature) prevents us from leveraging the fact that mappings are otherwise immutable. Eg. if an error occurs after a single MappedFieldTypeReference is updated, there is no way to go back. There are also a couple bugs I think, eg. after a mapping update that updates all types is executed, we should call refreshSource on all DocumentMapper instances so that this is reflected on their cached source.

In the end I suspect that the only way to make it sane really would be to remove the notion of types entirely. If we ever do that, we can just remove this method.